src/formio/components/FileField.js

```diff
@@ -212,9 +213,64 @@ class FileField extends Formio.Components.components.file {
       this.loading = false;
     });
 
+    let hasTypeErrors = false;
+    let statuses = [];
+
+    // Issue #3040 - Overwriting code from the super, in order to update the error message for wrong
+    // file types, so that it doesn't show the (escaped) MIME type to the user.
+    Array.prototype.forEach.call(files, file => {
+      const fileUpload = {
+        originalName: file.name,
+        name: FormioUtils.uniqueName(
+          file.name,
+          this.component.fileNameTemplate,
+          this.evalContext()
+        ),
+        size: file.size,
+        status: 'info',
+        message: this.t('Processing file. Please wait...'),
+      };
+
+      if (this.component.filePattern && !this.validatePattern(file, this.component.filePattern)) {
+        fileUpload.status = 'error';
+        fileUpload.message = this.t(
+          'The uploaded file is not of an allowed type. It must be: {{ pattern }}.',
+          {
+            pattern: this.formatAllowedTypesLabels(
+              this.component.file.allowedTypesLabels,
+              this.component.filePattern
+            ),
+          }
+        );
+        hasTypeErrors = true;
+      }
+
+      statuses.push(fileUpload);
+    });
+
+    if (hasTypeErrors) {
+      this.statuses = this.statuses.concat(statuses);
```

Comment:
We did push some objects with 'info' status on `statuses` But we only append them on `this.statuses` if an error has happened?

If that is correct, can we short circuit before the forEach if `this.component.filePattern` is falsy and `return super.upload(files)`?
And remove that invariant check from the callback function?

Reply:
I like the suggestion, but it's unclear to me how much of this code is ours and how much is copied from Form.io and modified. If it's mostly copied code from Form.io, I'd keep it as is so that it's easier to see the diff between ours and theirs so that we can eventually get rid of this again when better options present themselves (either a formio upgrade or our own formio-renderer).